subgraphs/mux-protocol/src/mappings/gainsHandlers.ts

```diff
@@ -0,0 +1,196 @@
+import {
+  BIGDECIMAL_ZERO,
+  BIGINT_ZERO,
+  DAI_ADDRESS_ARBITRUM,
+  GAINS_POOL_NAME,
+  GAINS_POOL_SYMBOL,
+  GAINS_PRECISION_DECIMALS,
+  BIGDECIMAL_HUNDRED,
+  GAINS_PAIRINFO_ADDRESS,
+  GAINS_VAULT_ADDRESS,
+  GAINS_REFERRALS_ADDRESS,
+  GAINS_MUX_REFERRER_ADDRESS,
+  INT_THREE,
+  INT_TWO,
+} from "../utils/constants";
+import { convertToDecimal, exponentToBigDecimal } from "../utils/numbers";
+import { RewardIntervalType, getRewardsPerDay } from "../entities/rewards";
+import { handleUpdatePositionEvent } from "./handlers";
+
+// Event emitted when a trade executes immediately, at the market price
+export function handleMarketExecuted(event: MarketExecuted): void {
+  const daiAddress = Address.fromString(DAI_ADDRESS_ARBITRUM);
+  let eventType = EventType.CollateralOut;
+  if (event.params.open) {
+    eventType = EventType.CollateralIn;
+  }
+  handleGainsUpdatePositionEvent(
+    event,
+    event.params.t.trader,
+    daiAddress,
+    event.params.positionSizeDai,
+    daiAddress,
+    event.params.positionSizeDai.times(event.params.t.leverage),
+    event.params.t.buy,
+    eventType,
+    event.params.percentProfit,
+    event.params.t.pairIndex
+  );
+}
+
+// Event emitted when a trade executes at exact price set if price reaches threshold
+export function handleLimitExecuted(event: LimitExecuted): void {
+  const daiAddress = Address.fromString(DAI_ADDRESS_ARBITRUM);
+  // orderType [TP, SL, LIQ, OPEN] (0-index)
+  let eventType = EventType.CollateralOut;
+  if (event.params.orderType == INT_THREE) {
+    eventType = EventType.CollateralIn;
+  } else if (event.params.orderType == INT_TWO) {
+    eventType = EventType.Liquidated;
+  }
+  handleGainsUpdatePositionEvent(
+    event,
+    event.params.t.trader,
+    daiAddress,
+    event.params.positionSizeDai,
+    daiAddress,
+    event.params.positionSizeDai.times(event.params.t.leverage),
+    event.params.t.buy,
+    eventType,
+    event.params.percentProfit,
+    event.params.t.pairIndex
+  );
+}
+
+export function handleGainsUpdatePositionEvent(
+  event: ethereum.Event,
+  accountAddress: Address,
+  collateralTokenAddress: Address,
+  collateralAmountDelta: BigInt,
+  indexTokenAddress: Address,
+  indexTokenAmountDelta: BigInt,
+  isLong: boolean,
+  eventType: EventType,
+  percentProfit: BigInt,
+  PairIndex: BigInt
+): void {
+  // If the referrer address from Gains Trade Referrals contract call is not related with MUX protocol, the trading does not come from MUX protocol.
+  const referralsContract = Referrals.bind(
+    Address.fromString(GAINS_REFERRALS_ADDRESS)
+  );
+  const referrer = referralsContract.getTraderReferrer(accountAddress);
+  if (referrer.toHexString().toLowerCase() != GAINS_MUX_REFERRER_ADDRESS) {
+    return;
+  }
+
+  const pool = getOrCreateLiquidityPool(
+    event,
+    Address.fromString(GAINS_VAULT_ADDRESS),
+    GAINS_POOL_NAME,
+    GAINS_POOL_SYMBOL
+  );
+  takeSnapshots(event, pool);
```

Comment:
You could potentially take the snapshot as part of the `getOrCreate`, so it is impossible to forget to take them accidentally

Reply:
Thank you very much for the suggestions. 

For this subgraph, I would like to keep it this way as there are limited event handlers to take snapshot. Meanwhile, as there may be several occasions to call getOrCreate() in an event handler, takeSnapshot() may be called several times in an event handler and as a result it may not be efficient enough if it is integrated into getOrCreate().